core/trino-main/src/main/java/io/trino/cost/HashPartitionCountProvider.java

```diff
@@ -0,0 +1,128 @@
+
+import static io.trino.SystemSessionProperties.getFaultTolerantExecutionPartitionCount;
+import static io.trino.SystemSessionProperties.getRemoteHashPartitionMinRowCount;
+import static io.trino.SystemSessionProperties.getRetryPolicy;
+import static io.trino.sql.planner.iterative.Lookup.noLookup;
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+import static java.util.Objects.requireNonNull;
+
+public final class HashPartitionCountProvider
+{
+    private HashPartitionCountProvider() {}
+
+    @VisibleForTesting
+    public static int getHashPartitionCount(Session session)
+    {
+        if (getRetryPolicy(session) == RetryPolicy.TASK) {
+            return getFaultTolerantExecutionPartitionCount(session);
+        }
+
+        return SystemSessionProperties.getHashPartitionCount(session);
+    }
+
+    public static int getHashPartitionCount(Session session, PlanNode node, StatsProvider statsProvider)
+    {
+        requireNonNull(session, "session is null");
+        requireNonNull(node, "node is null");
+        requireNonNull(statsProvider, "statsProvider is null");
+
+        if (getRetryPolicy(session) == RetryPolicy.TASK) {
+            return getFaultTolerantExecutionPartitionCount(session);
+        }
+
+        return getHashPartitionCount(
+                node,
+                statsProvider,
+                SystemSessionProperties.getHashPartitionCount(session),
+                getRemoteHashPartitionMinRowCount(session));
+    }
+
+    private static int getHashPartitionCount(
+            PlanNode root,
+            StatsProvider statsProvider,
+            int maxHashPartitionCount,
+            long remoteHashPartitionMinRowCount)
+    {
+        if (remoteHashPartitionMinRowCount < 1) {
+            return maxHashPartitionCount;
+        }
+
+        if (isExpandingNodesPresent(root)) {
+            return maxHashPartitionCount;
+        }
+
+        double sourceTablesRowCount = getSourceTablesRowCount(root, statsProvider);
+        double maxJoinRowCount = getMaxJoinRowCount(root, statsProvider);
+        if (Double.isNaN(sourceTablesRowCount) || Double.isNaN(maxJoinRowCount)) {
+            return maxHashPartitionCount;
+        }
+
+        int hashPartitionCount = max(
+                getHashPartitionsFromRowCount(sourceTablesRowCount, remoteHashPartitionMinRowCount),
+                getHashPartitionsFromRowCount(maxJoinRowCount, remoteHashPartitionMinRowCount));
+        return min(hashPartitionCount, maxHashPartitionCount);
+    }
+
+    private static int getHashPartitionsFromRowCount(double rowCount, long remoteHashPartitionMinRowCount)
+    {
+        return max((int) (rowCount / remoteHashPartitionMinRowCount), 1);
+    }
+
+    private static boolean isExpandingNodesPresent(PlanNode root)
+    {
+        return PlanNodeSearcher.searchFrom(root, noLookup())
+                .where(node -> ((node instanceof JoinNode) && ((JoinNode) node).isCrossJoin()) || (node instanceof UnnestNode))
```

Comment:
not every cross join is expanding. For example, planner will insert cross join deliberately (`PushAggregationThroughOuterJoin`) where one side has single row